R/utilities-units.R

```diff
@@ -336,3 +336,200 @@ initializeDimensionAndUnitLists <- function() {
+#'   `ospsuite::ospUnits`, respectively.
+#'
+#' @seealso toUnit
+#'
+#' @examples
+#'
+#' # small dataframe to illustrate the conversion
+#' (df <- dplyr::tibble(
+#'   dataType = c(rep("simulated", 3), rep("observed", 3)),
+#'   xValues = c(0, 14.482, 28.965, 0, 1, 2),
+#'   xUnit = "min", xDimension = "Time",
+#'   yValues = c(1, 1, 1, 1, 1, 1),
+#'   yUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   yDimension = c("Amount", "Amount", "Amount", "Mass", "Mass", "Mass"),
+#'   yErrorValues = c(2.747, 2.918, 2.746, NA, NA, NA),
+#'   yErrorUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   molWeight = c(10, 10, 20, 20, 20, 10)
+#' ))
+#'
+#' # default conversion
+#' ospsuite:::.unitConverter(df)
+#'
+#' # customizing conversion with specified unit(s)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h)
+#' ospsuite:::.unitConverter(df, yUnit = ospUnits$Mass$kg)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$s, yUnit = ospUnits$Amount$mmol)
+#'
+#' @keywords internal
+.unitConverter <- function(data, xUnit = NULL, yUnit = NULL) {
+  # target units --------------------------
+
+  # No validation of inputs for this non-exported function.
+  # All validation will take place in the `DataCombined` class itself.
+
+  # The observed and simulated data should have the same units for
+  # visual/graphical comparison.
+  #
+  # Therefore, if target units are not specified by the user, we need to choose
+  # one ourselves. For no special reason, the first element from a vector of
+  # unique units will be selected: one for X-axis, and one for Y-axis, i.e.
+  xTargetUnit <- xUnit %||% unique(data$xUnit)[[1]]
+  yTargetUnit <- yUnit %||% unique(data$yUnit)[[1]]
+
+  # internal --------------------------
+
+  # The strategy is to split the data frame for each source unit and carry out
+  # conversion separately per data frame. This is the most performant options
+  # since there can only be as many expensive calls to `toUnit()`/`rClr` as
+  # there are source units.
+  #
+  # The problem occurs when source units are missing (`NA`). The `toUnit`
+  # function can handle them but not `split()`, which would drop the entire
+  # section of the data frame corresponding to `NA` and thus there will be loss
+  # of data when a data frame is split into a list of data frames.
+  #
+  # The trick is to create copies of source unit and molecular weight columns
+  # and fill in the missing values with something other than `NA`. Using these
+  # new columns with `split()` makes sure that parts of a data frame where
+  # source units are missing won't be dropped. Note that the original columns
+  # containing source units remain untouched.
+  #
+  # These newly created columns are removed before the converted data frame is
+  # returned to the user.
+  data <- data %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Unit$|Weight$"),
+        .fns = as.character,
+        .names = "{.col}Split"
```

Comment:
Is this going to add Split at the end of the name for the one matching?

I feel that almost every line of code needs a comment.

Reply:
> Is this going to add Split at the end of the name for the one matching?

Yes.

> I feel that almost every line of code needs a comment.

Done. They are the worst kind of comments: the ones that just reiterate the code. But, I guess it is essential in this case, since not every contributor can understand that code easily.

Reply:
>They are the worst kind of comments: the ones that just reiterate the code

This not what I am asking here. Do not comment the obvious


```
# Return data
return(data) 
```
is indeed not good.

Comment that is going to explain some 3rd party code, that no one in the team is familiar with, is in my opinion not only good, it's absolutely required so that we can maintain this in the future. Until you have managed to bring the team to your level, this is a requirement.

Reply:
Okay, fair enough. I had added the comments, but, for the sake of completeness, I've also made a PR to update coding standards to reflect this: https://github.com/Open-Systems-Pharmacology/Suite/pull/108

Also, I want to mention that `DataCombined` code also contains such clarifying comments, which were indeed included to make it easier for the team to understand the code.